### ISSUE TITLE
Array to string conversion in Zend\Log\Writer\Db

### ISSUE THREAD

**Issue by Martin-P** [zendframework/zf2#7244]
This was already reported in #2589, but that issue got closed 11 days ago (https://github.com/zendframework/zf2/issues/2589#issuecomment-73774768) with references to:
- PR #5254 which is about a db adapter via config and is unrelated to the issue
- Issue #5603 which is a duplicate of the issue

I tried to get it reopened, but without any success.

Testcase which shows the problem:
```php
/**
 * @group 2589
 */
public function testMapEventIntoColumnDoesNotTriggerArrayToStringConversion()
{
    $this->writer = new DbWriter($this->db, $this->tableName, array(
        'priority' => 'new-priority-field',
        'message'  => 'new-message-field' ,
        'extra'    => array(
            'file'  => 'new-file',
            'line'  => 'new-line',
            'trace' => 'new-trace',
        )
    ));

    // log to the mock db adapter
    $priority = 2;
    $message  = 'message-to-log';
    $extra    = array(
        'file'  => 'test.php',
        'line'  => 1,
        'trace' => array(
            array(
                'function' => 'Bar',
                'class'    => 'Foo',
                'type'     => '->',
                'args'     => array(
                    'baz',
                ),
            ),
        ),
    );
    $this->writer->write(array(
        'priority' => $priority,
        'message'  => $message,
        'extra'    => $extra,
    ));

    $this->assertContains('query', array_keys($this->db->calls));
    $this->assertEquals(1, count($this->db->calls['query']));

    foreach ($this->db->calls['execute'][0][0] as $fieldName => $fieldValue) {
        $this->assertInternalType('string', $fieldName);
        $this->assertInternalType('string', (string) $fieldValue);
    }
}
```

Result:
> 1) ZendTest\Log\Writer\DbTest::testMapEventIntoColumnDoesNotTriggerArrayToStringConversion
Array to string conversion

**Comment by weierophinney** [comment]
@Martin-P Are you working on a PR for this, by any chance?

**Comment by Martin-P** [comment]
@weierophinney 
Looking at it right now, but there are some issues:
- `serialize()`: not an option, because it can't handle closures
- `var_export()`: not an option, because it will create a huge string which leads to situations as in #2592 (every class with all its properties is exported)
- `print_r()`: not an option, because it also creates a huge string where all (nested) classes and properties are exported
- `json_encode()`: works, but objects are stored as stdClass and this causes loss of data

I think a custom routine is the best approach to do this. The main issues are with the depth of the array and size due to possible objects inside the array. I will create a PR for this with a suggested fix.

**Comment by Martin-P** [comment]
I think I will go for `var_export()` after all. My use case which creates a huge trace is when `Zend\Log\Writer\Db` has to save an `Exception`, but I think it is better to format the `Exception` correctly in `Zend\Log\Formatter\ExceptionHandler`.

**Comment by Martin-P** [comment]
I created #7282 for this.

**Comment by weierophinney** [comment]
Fixed with #7282.
